File: hw/fuentes/U3/HW22STRATEGY/SortAppPY/ec/edu/espe/clientApp/model/quick_sort.py
from model.sorting_strategy import SortingStrategy

class QuickSort(SortingStrategy):
    def sort(self, data: list[int]) -> list[int]:
        result = data.copy()
        self._quick_sort(result, 0, len(result) - 1)
        print("Using QuickSort")
        return result

    def _quick_sort(self, arr: list[int], low: int, high: int):
        if low < high:
            pi = self._partition(arr, low, high)
            self._quick_sort(arr, low, pi - 1)
            self._quick_sort(arr, pi + 1, high)

    def _partition(self, arr: list[int], low: int, high: int) -> int:
        pivot = arr[high]
        i = low - 1
        for j in range(low, high):
            if arr[j] < pivot:
                i += 1
                arr[i], arr[j] = arr[j], arr[i]
        arr[i + 1], arr[high] = arr[high], arr[i + 1]
        return i + 1

File: exams/jacome/U3/ExamU3/model/QuickSort.js
import { SortingStrategy } from './SortingStrategy.js';

export class QuickSort extends SortingStrategy {
  sort(arr) {
    if (arr.length <= 1) return arr;
    const pivot = arr[Math.floor(arr.length / 2)];
    const left = arr.filter(x => x < pivot);
    const middle = arr.filter(x => x === pivot);
    const right = arr.filter(x => x > pivot);
    return this.sort(left).concat(middle, this.sort(right));
  }
}
